File: automata_builder/renderer.py
from __future__ import annotations

import copy
from typing import Sequence


class RenderFrame(object):
    def __init__(self, lines: Sequence[str] = ()):
        lengths = [len(line) for line in lines]
        assert len(set(lengths)) <= 1, "All lines must have the same length"
        self.lines = list(lines)[:]

    def get_lines(self) -> list[str]:
        return self.lines[:]

    def get_width(self) -> int:
        if not self.lines:
            return 0

        return len(self.lines[0])

    def get_height(self) -> int:
        return len(self.lines)

    def get_dimensions(self):
        return self.get_height(), self.get_width()

    @classmethod
    def from_line(cls, line: str) -> RenderFrame:
        return RenderFrame([line])

    def render(self) -> str:
        return '\n'.join(self.lines)

    def __repr__(self):
        name = self.__class__.__name__
        return f'{name}(lines={self.lines})'

    def add_line(self, line: str) -> None:
        if len(line) != self.get_width():
            raise ValueError("Line length must match frame width")

        self.lines.append(line)

    def extend_down(self, other: RenderFrame) -> RenderFrame:
        if self.get_height() == 0:
            self.lines = other.get_lines()
            return self

        if self.get_width() != other.get_width():
            print(self.get_width(), other.get_width())
            raise ValueError("Frame widths must match to extend down")

        self.lines.extend(other.lines)
        return self

    def extend_up(self, other: RenderFrame) -> RenderFrame:
        if self.get_height() == 0:
            self.lines = other.get_lines()
            return self

        if self.get_width() != other.get_width():
            raise ValueError("Frame widths must match to extend up")

        self.lines = other.lines + self.lines
        return self

    def extend_left(self, other: RenderFrame) -> RenderFrame:
        if self.get_height() == 0:
            self.lines = other.get_lines()
            return self

        if self.get_height() != other.get_height():
            raise ValueError("Frame heights must match to extend left")

        new_lines = []
        for line_self, line_other in zip(self.get_lines(), other.get_lines()):
            new_lines.append(line_other + line_self)

        self.lines = new_lines
        return self

    def extend_right(self, other: RenderFrame) -> RenderFrame:
        if self.get_height() == 0:
            self.lines = other.get_lines()
            return self

        if self.get_height() != other.get_height():
            raise ValueError("Frame heights must match to extend right")

        new_lines = []
        for line_self, line_other in zip(self.get_lines(), other.get_lines()):
            new_lines.append(line_self + line_other)

        self.lines = new_lines
        return self

    @classmethod
    def join_vertically(cls, frames: list[RenderFrame]) -> RenderFrame:
        combined_frame = RenderFrame()
        for frame in frames:
            combined_frame.extend_down(frame)

        return combined_frame

    @classmethod
    def join_horizontally(cls, frames: list[RenderFrame]) -> RenderFrame:
        combined_frame = RenderFrame()
        for frame in frames:
            combined_frame.extend_right(frame)

        # print("Combined frame lines:", combined_frame.get_lines())
        return combined_frame

File: src/automata/renderer.rs
use std::fmt;
use std::ops::{Deref, DerefMut};

#[derive(Debug, Clone, Default, Eq, PartialEq)]
pub struct RenderFrame {
    lines: Vec<String>,
}
impl RenderFrame {
    /// All lines must have the same length, otherwise `None` is returned.
    pub fn new(lines: Vec<String>) -> Option<RenderFrame> {
        let first_len = lines.first().map(|line| line.chars().count());
        if let Some(len) = first_len {
            if lines.iter().any(|line| line.chars().count() != len) {
                return None;
            }
        }
        Some(RenderFrame { lines })
    }
    /// Pads every line on the right so that all of them share the widest length.
    pub fn from_padded_lines(lines: Vec<String>) -> RenderFrame {
        let width = lines.iter().map(|line| line.chars().count()).max().unwrap_or(0);
        let padded = lines
            .into_iter()
            .map(|line| {
                let pad = width - line.chars().count();
                line + &" ".repeat(pad)
            })
            .collect();
        RenderFrame { lines: padded }
    }
    pub fn from_line(line: String) -> RenderFrame {
        RenderFrame { lines: vec![line] }
    }
    pub fn get_lines(&self) -> &[String] {
        &self.lines
    }
    pub fn get_width(&self) -> usize {
        self.lines.first().map(|line| line.chars().count()).unwrap_or(0)
    }
    pub fn get_height(&self) -> usize {
        self.lines.len()
    }
    pub fn get_dimensions(&self) -> (usize, usize) {
        (self.get_height(), self.get_width())
    }
    pub fn render(&self) -> String {
        self.lines.join("\n")
    }

    /// Appends `other`'s lines below our own. Widths must match
    /// unless the current frame is still empty.
    pub fn extend_down(&mut self, other: &RenderFrame) -> Result<&mut Self, RenderError> {
        if self.get_height() == 0 {
            self.lines = other.lines.clone();
            return Ok(self);
        }
        if self.get_width() != other.get_width() {
            return Err(RenderError::WidthMismatch {
                own: self.get_width(),
                other: other.get_width(),
            });
        }
        self.lines.extend(other.lines.iter().cloned());
        Ok(self)
    }

    /// Appends `other`'s lines to the right of our own. Heights must match
    /// unless the current frame is still empty.
    pub fn extend_right(&mut self, other: &RenderFrame) -> Result<&mut Self, RenderError> {
        if self.get_height() == 0 {
            self.lines = other.lines.clone();
            return Ok(self);
        }
        if self.get_height() != other.get_height() {
            return Err(RenderError::HeightMismatch {
                own: self.get_height(),
                other: other.get_height(),
            });
        }
        for (own_line, other_line) in self.lines.iter_mut().zip(other.lines.iter()) {
            own_line.push_str(other_line);
        }
        Ok(self)
    }

    pub fn join_vertically(frames: &[RenderFrame]) -> Result<RenderFrame, RenderError> {
        let mut combined = RenderFrame::default();
        for frame in frames {
            combined.extend_down(frame)?;
        }
        Ok(combined)
    }

    pub fn join_horizontally(frames: &[RenderFrame]) -> Result<RenderFrame, RenderError> {
        let mut combined = RenderFrame::default();
        for frame in frames {
            combined.extend_right(frame)?;
        }
        Ok(combined)
    }
}

/// Equivalent of the `TapeRenderFrame` subclass.
/// Rust has no inheritance, so this wraps a `RenderFrame` and exposes its
/// methods through `Deref`/`DerefMut`.
pub struct TapeRenderFrame {
    frame: RenderFrame,
    pub(crate) num_cells: usize,
    cell_width: usize,
}

impl TapeRenderFrame {
    pub fn new(line: &str, num_cells: usize, cell_width: usize) -> Self {
        TapeRenderFrame {
            frame: RenderFrame::from_line(line.parse().unwrap()),
            num_cells,
            cell_width,
        }
    }

    pub fn get_space_consumed(&self) -> usize {
        self.num_cells * (self.cell_width + 1)
    }
}

impl Deref for TapeRenderFrame {
    type Target = RenderFrame;

    fn deref(&self) -> &RenderFrame {
        &self.frame
    }
}

impl DerefMut for TapeRenderFrame {
    fn deref_mut(&mut self) -> &mut RenderFrame {
        &mut self.frame
    }
}

impl fmt::Debug for TapeRenderFrame {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "TapeRenderFrame(lines={:?}, num_cells={}, cell_width={})",
            self.frame.lines, self.num_cells, self.cell_width
        )
    }
}


#[derive(Debug, Clone, Eq, PartialEq)]
pub enum RenderError {
    WidthMismatch { own: usize, other: usize },
    HeightMismatch { own: usize, other: usize },
}
impl std::fmt::Display for RenderError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            RenderError::WidthMismatch { own, other } => {
                write!(f, "Frame widths must match: {} vs {}", own, other)
            }
            RenderError::HeightMismatch { own, other } => {
                write!(f, "Frame heights must match: {} vs {}", own, other)
            }
        }
    }
}
impl std::error::Error for RenderError {}
